lib/pure/collections/lists.nim

```diff
@@ -7,24 +7,148 @@
 #    distribution, for details about the copyright.
 #
 
-## Implementation of singly and doubly linked lists. Because it makes no sense
-## to do so, the 'next' and 'prev' pointers are not hidden from you and can
+## Implementation of singly and doubly linked lists and rings.
+##
+## Because it makes no sense to do so, the ``next`` and ``prev``
+## pointers are not hidden from you and can
 ## be manipulated directly for efficiency.
+##
+## * ``DoublyLinkedList`` and ``DoublyLinkedRing`` are ``O(1)`` to `remove` elements
+## * ``SinglyLinkedList`` and ``SingleLinkedRing`` are ``O(n)`` to `remove` elements `(with lower memory usage)`
+##
+## Initializing a list/ring
+## ====================
+##
+## .. code-block:: Nim
+##   import lists, random
+##   var
+##     emptyIntDlr = initDoublyLinkedRing[int]()
+##     emptyStrSll = initSinglyLinkedList[string]()
+##     sll = @[1, 2, 3, 4].toSinglyLinkedList()
+##     dll = @[1, 2, 3, 4].toDoublyLinkedList()
+##     slr = @[1, 2, 3, 4].toSinglyLinkedRing()
+##     dlr = @[1, 2, 3, 4].toDoublyLinkedRing()
+##     randDll = newDoublyLinkedListWith(10, random(100))  # dll of 10 random numbers
+##   echo randDll
+##
+## Converting a list/ring
+## ======================
+##
+## .. code-block:: nim
+##   import lists
+##   var
+##     sll = @[1, 2, 3, 4].toSinglyLinkedList()
+##     sqSll = sll.toSeq
+##     dll = @[1, 2, 3, 4].toDoublyLinkedList()
+##     sqDll = dll.toSeq
+##     slr = @[1, 2, 3, 4].toSinglyLinkedRing()
+##     sqSlr = slr.toSeq
+##     dlr = @[1, 2, 3, 4].toDoublyLinkedRing()
+##     sqDlr = dlr.toSeq
+##
+## Transforming a list/ring
+## ========================
+##
+## .. code-block:: nim
+##   import lists
+##   var
+##     sll1 = @[1, 2, 3, 4].toSinglyLinkedList()
+##     sll2 = map(sll1, proc(x: int): string = $x)  # MAP
+##     dll1 = @[1, 2, 3, 4].toDoublyLinkedList()
+##     dll2 = map(dll1, proc(x: int): string = $x)  # MAP
+##     slr1 = @[1, 2, 3, 4].toSinglyLinkedRing()
+##     slr2 = map(slr1, proc(x: int): string = $x)  # MAP
+##     dlr1 = @[1, 2, 3, 4].toDoublyLinkedRing()
+##     dlr2 = map(dlr1, proc(x: int): string = $x)  # MAP
+##
+##   assert sll2 == @["1", "2", "3", "4"].toSinglyLinkedList()
+##   assert dll2 == @["1", "2", "3", "4"].toDoublyLinkedList()
+##   assert slr2 == @["1", "2", "3", "4"].toSinglyLinkedRing()
+##   assert dlr2 == @["1", "2", "3", "4"].toDoublyLinkedRing()
+##
+##   var sll = @["1", "2", "3", "4"].toSinglyLinkedList()
+##   apply(sll, proc(x: var string) = x &= "42")  # APPLY #1
+##   echo sll   # --> ["142", "242", "342", "442"]
+##
+##   var dll = @["1", "2", "3", "4"].toSinglyLinkedList()
```

Comment:
Especially the following bunch of examples though, they seems completely redundant. It's the exact same code!